File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        result = []
        def smallest(node):
            if not node: return
            smallest(node.left)
            
            if len(result) == k:
                return
                
            result.append(node.val)
            smallest(node.right)
        smallest(root)
        return result[-1]
        

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int count = 0;
        while (root != null) {
            if (root.left == null) {
                count++;
                if (count == k) {
                    return root.val;
                }
                root = root.right;
            } else {
                TreeNode predecessor = root.left;
                while (predecessor.right != null && predecessor.right != root) {
                    predecessor = predecessor.right;
                }
                if (predecessor.right == null) {
                    predecessor.right = root;
                    root = root.left;
                } else {
                    predecessor.right = null;
                    count++;
                    if (count == k) {
                        return root.val;
                    }
                    root = root.right;
                }
            }
        }
        return -1; // Not found if k is invalid
    }
}
